README-permission.md

```diff
@@ -0,0 +1,163 @@
+  - name: Ensure permission MyPermission is present with the User Administrators privilege present
+    ipapermission:
+      ipaadmin_password: SomeADMINpassword
+      name: MyPermission
+      privilege: "User Administrators"
+      action: member
+```
+
+
+Example playbook to make sure permission "MyPermission" member "privilege" with value "User Administrators" is absent:
+
+
+```yaml
+---
+- name: Permission remove privilege from a permission
+  hosts: ipaserver
+  become: true
+
+  tasks:
+  - name: Ensure permission MyPermission is present without the User Administrators privilege
+    ipapermission:
+      ipaadmin_password: SomeADMINpassword
+      name: MyPermission
+      privilege: "User Administrators"
+      action: member
+      state: absent
+```
+
+
+Example playbook to make sure permission "MyPermission" is absent:
+
+```yaml
+---
+- name: Playbook to manage IPA permission.
+  hosts: ipaserver
+  become: yes
+
+  tasks:
+  - ipapermission:
+      ipaadmin_password: SomeADMINpassword
+      name: MyPermission
+      state: absent
+```
+
+Example playbook to make sure permission "MyPermission" is renamed to "MyNewPermission":
+
+```yaml
+---
+- name: Playbook to manage IPA permission.
+  hosts: ipaserver
+  become: yes
+
+  tasks:
+  - ipapermission:
+      ipaadmin_password: SomeADMINpassword
+      name: MyPermission
+      rename: MyNewPermission
+      state: renamed
+```
+
+
+
+
+Variables
+---------
+
+ipapermission
+-------
+
+Variable | Description | Required
+-------- | ----------- | --------
+`ipaadmin_principal` | The admin principal is a string and defaults to `admin` | no
+`ipaadmin_password` | The admin password is a string and is required if there is no admin ticket available on the node | no
+`name` \| `cn` | The permission name string. | yes
+`perm_rights` \| `ipapermright` | Rights to grant. It can be a list of one or more of `read`, `search`, `compare`, `write`, `add`, `delete`, and `all` default: `all` | no
+`attrs` | All attributes to which the permission applies | no
+`perm_bind_rule_type` \| `ipapermbindruletype` | Bind rule type. It can be one of `permission`, `all`, `self`, or `anonymous` defaults to `permission` for new permissions.| no
+`perm_location` \| `ipapermlocation` | Subtree to apply permissions to | no
+`extra_target_filter` \| `extratargetfilter` | Extra target filter | no
+`perm_target_filter` \| `ipapermtargetfilter` | All target filters | no
+`perm_target` \| `ipapermtarget` | Optional DN to apply the permission to | no
+`perm_target_to` \| `ipapermtargetto` | Optional DN subtree where an entry can be moved to | no
+`perm_target_from` \| `ipapermtargetfrom` | Optional DN subtree from where an entry can be moved | no
+`member_of` \| `memberof` | Target members of a group (sets memberOf targetfilter) | no
+`target_group` \| `targetgroup` | User group to apply permissions to (sets target) | no
+`object_type` \| `type` | Type of IPA object (sets subtree and objectClass targetfilter) | no
```

Comment:
Variables should follow the names on CLI, with aliases for the API. So, for example `perm_rights` should be `right` (and not `right_s_`, as it is another parameter on the CLI), `perm_bind_rule_type` should be `bindtype`, and so on.

I also didn't find a `bindtype` with value `self`. Is it for `selfservice`? Does it depend on IPA version?

Reply:
I can make the changes to the parameter names. The logic makes sense. I was just making things up.

As for the perm_bind_rule_type of self, I don't see it in on IPA version 4.6.8 from our RHEL 7 instance, but the FreeIPA Demo instance running 4.8.7 has it in their api doc. I included it for future proofing. 

I guess that would limit the version compatibility, but everything else I have tried seems to work with 4.6.8. If you would like, I can put a note in the version compatibility part of the documentation.

Reply:
The `self` value is only supported on IPA 4.8.7 and later.

For now, don't do anything in the code. It is fine. You only need to "protect" the tests that use it in a block with a condition to only execute them if the right IPA version is found. Check `tests/hostgroup/test_hostgroup_rename.yml` to see how it is handled.

Reply:
It would be really good to have an error message in this case. Something like:
`The self value is not  supported by your IPA version`

Reply:
@rjeffman Sounds good. I will get on it. I will try not to disapear for 2 weeks this time. 

@t-woerner Is there a good example of how to determine the IPA version in a module? I feel like this should/could be a function for this in `plugins/module_utils/ansible_freeipa_module.py`, but I could also call the API, using the existing API functions, in the permissions module.

Reply:
@kresss, there is, currently, no way to check if a value for a parameter is available or not.

Take a look at PR #419 where I added a function to check for the FreeIPA version, so you can use it to verify if the version you want have the proper value. For `self`, you want to test against version `4.8.7`, and if the version fails, you should fail with a message like `"bindtype 'self' is not supported by your IPA version."`

You should also add a test to verify this behavior:

```yaml
- name: Test bindtype `self`availability
  ipapermission:
    name: test_self
    bindtype: self
  register: result
  failed_when: not result.failed or "bindtype 'self' is not supported by your IPA version." not in result.msg
  when: ipa_version is version('4.8.7', '<')
```

Note that this test will only fail if ipa_version is less then 4.8.7. To have `ipa_version` available in the test playbook, you have to include `env_freeipa_facts.yml` in the begining:
```yaml
- include_tasks: ../env_freeipa_facts.yml
```